src/test/pages/EventDetails.test.tsx

```diff
@@ -0,0 +1,230 @@
+      role: role,
+      extraEndpoints: [
+        rest.get(API_PATH + "/events", (req, res, ctx) => {
+          return res(ctx.status(200), ctx.json({ results: [event], totalResults: 0 }));
+        }),
+        rest.get(API_PATH + "/events/:eventId", (req, res, ctx) => {
+          return res(ctx.status(200), ctx.json(event));
+        }),
+      ],
+    });
+    const eventCards = await screen.findAllByTestId("event-card");
+    const eventCardViewDetails = within(eventCards[0]).getByText("View details");
+    await userEvent.click(eventCardViewDetails);
+  };
+
+  it("renders without crashing", async () => {
+    await setupTest({ role: "STUDENT", event: mockEvent as IsaacEventPageDTO });
+    checkPageTitle(mockEvent.title);
+  });
+
+  it("displays event title a second time with image, if image is defined", async () => {
+    await setupTest({ role: "STUDENT", event: mockEvent as IsaacEventPageDTO });
+    const { title, image } = getFields();
+    expect(title).toBeInTheDocument();
+    expect(image).toHaveAttribute("src", mockEvent.eventThumbnail.src);
+  });
+
+  it("uses a placeholder image if no image is specified", async () => {
+    await setupTest({ role: "STUDENT", event: { ...mockEvent, eventThumbnail: undefined } as IsaacEventPageDTO });
+    const { image } = getFields();
+    expect(image).toHaveAttribute("src", "http://placehold.it/500x276");
+  });
+
+  it("shows a map and address details if location, latitude and longitude are specified, and event is not tagged as online", async () => {
+    const mapLocation = {
+      address: { addressLine1: "Mock street", town: "Mock town", postalCode: "ABC 123" },
+      latitude: 53.95038292218263,
+      longitude: -1.0510007751299424,
+    };
+    await setupTest({
+      role: "STUDENT",
+      event: { ...mockEvent, tags: ["booster", "student"], location: mapLocation } as IsaacEventPageDTO,
+    });
+    const { map, location } = getFields();
+    expect(map).toBeInTheDocument();
+    expect(location).toHaveTextContent(formatAddress(mapLocation));
+  });
+
+  it("shows location as online if event is tagged as virtual", async () => {
+    await setupTest({ role: "STUDENT", event: { ...mockEvent, tags: ["virtual"] } as IsaacEventPageDTO });
+    const { location } = getFields();
+    expect(location).toHaveTextContent("Online");
+  });
+
+  it("shows private event badge if event is private", async () => {
+    await setupTest({ role: "STUDENT", event: { ...mockEvent, privateEvent: true } as IsaacEventPageDTO });
+    const { privateBadge } = getFields();
+    expect(privateBadge).toBeVisible();
+  });
+
+  it("does not show private event badge if even is not private", async () => {
+    await setupTest({ role: "STUDENT", event: { ...mockEvent, privateEvent: false } as IsaacEventPageDTO });
+    const { privateBadge } = getFields();
+    expect(privateBadge).not.toBeInTheDocument();
+  });
+
+  it.each(["ADMIN", "EVENT_MANAGER", "CONTENT_EDITOR"] as TestUserRole[])(
```

Comment:
Small one, but this can simply be `UserRole[]`.

Reply:
`type TestUserRole = UserRole | "ANONYMOUS"`

`TestUserRole[]` is what I use anytime there might be "ANONYMOUS" (which is not a user role on the platform but is what we use for the test environment to have a not-logged-in user).  The alternative is writing `UserRole[] | "ANONYMOUS"` each time.

On this particular line I am only testing 3 User Roles, but elsewhere in the same file I have to use `TestUserRole[]` for "ANONYMOUS" so I figured it made more sense to just use the same one for consistency and avoid importing multiple types.

Reply:
I understand, but importing multiple times is not a problem. Defining this as `UserRole[]` has the advantage of clearly stating that you don't care about anonymous users in this test. It is important to always define the right type for every variable, constant, parameter, return type, etc. as it helps both the developers reading your code and the compiler to make optimisations, suggestions and find errors. It's good practice. Previous developers didn't care too much, and now we have a codebase full of `any`s...